PSA: Phase 1: Add PK_OPAQUE_PSA type

mpg opened the pull request:
This implements a new PK type that can be used to wrap ECDSA keys held by PSA, and perform signature with them, plus a few administrative queries.

Remaining work / Open questions:

- ~~need to deduplicate function and loop copied from https://github.com/ARMmbed/mbedtls/pull/2159~~ fixed by rebasing on #2206  
- ~~should probably align names with https://github.com/ARMmbed/mbedtls/pull/2156 - opaque, psa, opaque_psa (at least for public API functions)~~ done

**Note:** based on #2206, only newer commits need reviewing.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -716,4 +720,214 @@ const mbedtls_pk_info_t mbedtls_rsa_alt_info = {
+    void *ctx = mbedtls_calloc( 1, sizeof( psa_key_slot_t ) );
+
+    if( ctx != NULL )
+        memset( ctx, 0, sizeof( psa_key_slot_t ) );
```
Side-question: Standard `calloc()` clears the content of the newly allocated buffer - do we require that for potential replacements, too?

mpg replied:
I think we do. Tbh, I'm not entirely sure why we have that code block here - I think it's for uniformity with the case where we have an `_init()` function to call, but I'm not sure that's a good reason. Wdyt?

mpg replied:
Replacing that call with a comment.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -716,4 +720,214 @@ const mbedtls_pk_info_t mbedtls_rsa_alt_info = {
+    *p -= len;
+    memcpy( *p, src, len );
+
+    if( **p & 0x80 )
```
This is only valid for `len != 0`, which might be treated independently.

hanno-arm replied:
That's also a bug in `mbedtls_asn1_write_mpi()` I think: #2166.

mpg replied:
Since this is a static function, and neither r not s can be zero in an ECDSA signature, I don't think we need to care for `len == 0` here, but it's certainly worth adding the point I just made as a comment.

mpg replied:
Adding the comment.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -716,4 +720,214 @@ const mbedtls_pk_info_t mbedtls_rsa_alt_info = {
+                                       unsigned char *dst )
+{
+    int ret;
+    unsigned char buf[MBEDTLS_ECDSA_MAX_LEN];
```
I'd prefer to use a signature which additionally has a `size_t dst_len` for the size of the total size of the destination buffer, and a `size_t* olen` to report the actual size.

mpg replied:
This was changed singificantly while getting rid of the stack buffers.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -716,4 +720,214 @@ const mbedtls_pk_info_t mbedtls_rsa_alt_info = {
+                                       unsigned char *dst )
+{
+    int ret;
+    unsigned char buf[MBEDTLS_ECDSA_MAX_LEN];
```
I think we should avoid such large stack allocations and work in place on the target buffer.

mpg replied:
In principle I would agree, but unfortunately we don't know the size of the output buffer, so it we want to use it directly we can't write to its end and then to a `memmove()`, we'd have to first compute the length of the signature, then write it starting from its end. It would make the code more complex and harder to maintain (need to keep in sync the length computation with the ASN.1 writing code).

Also, please note that so far we already had such a buffer in `ecdsa_signature_to_asn1()`, which is not used by `psa_ecdsa_sign()`, so compared to the non-PSA way, the increase of stack usage (disregarding frames and small local variables) is from `MBEDTLS_ECDSA_MAX_LEN` to `MBEDTLS_ECDSA_MAX_LEN + 2 * MBEDTLS_ECP_MAX_BYTES`. So an alternative to what I wrote above would allocate a single buffer of size `MBEDTLS_ECDSA_MAX_LEN` in `pk_psa_sign_wrap()` and then make `pk_ecdsa_sig_asn1_from_psa` and still `memcpy` to the actual destination buffer at the end. This would avoid the complexity of computing length before writing, and keep usage of large stack buffers in line with the non-PSA case.

Please let me know what you think.

hanno-arm replied:
> In principle I would agree, but unfortunately we don't know the size of the output buffer, so it we want to use it directly we can't write to its end and then to a memmove(), we'd have to first compute the length of the signature, then write it starting from its end. It would make the code more complex and harder to maintain (need to keep in sync the length computation with the ASN.1 writing code).

I agree it's a bit tedious, but I think it's worth it. Also, I think one could even get rid of _any_ stack buffers by calling `psa_ecdsa_sign()` on the destination buffer (which we assume is large enough to hold the ASN.1 version of the signature, and so it must also be large enough for the raw one), and then modify `ecdsa_signature_to_asn1()` to work in place.

mpg replied:
I got rid of both stack buffers. My solution is based on pre-computing the maximal length of any signature for this curve size, rather than the actual size of this particular signature, so there's a `memmove()` call near the end of `pk_ecdsa_sig_asn1_from_psa()` that could be avoided, but I think it's an acceptable compromise.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -716,4 +720,214 @@ const mbedtls_pk_info_t mbedtls_rsa_alt_info = {
+    const psa_key_slot_t *key = (const psa_key_slot_t *) ctx;
+    psa_status_t status;
+    psa_algorithm_t alg = PSA_ALG_ECDSA( translate_md_to_psa( md_alg ) );
+    /* PSA needs a buffer of know size */
```
Typo: `know` -> `known`

mpg replied:
Fixed.

hanno-arm commented on tests/suites/test_suite_pk.function:
```diff
@@ -90,13 +90,128 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
+
+    /* find a free key slot
+     * this is temporary, a new PSA API for this is coming */
+    for( key = 1; key <= 32; key++ )
```
Just a reminder that this too should be shared with @AndrzejKurek's code from #2159.

mpg replied:
Will be addressed when rebasing on #2184.

hanno-arm commented on tests/suites/test_suite_pk.function:
```diff
@@ -90,13 +90,128 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
+/* BEGIN_CASE depends_on:MBEDTLS_USE_PSA_CRYPTO:MBEDTLS_ECDSA_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED */
+void pk_psa_utils(  )
+{
+    mbedtls_pk_context pk, pk2;
```
We should add a call to `psa_init()` somewhere. Probably in the [main() in main.function](https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/main_test.function#L216) next to the platform initialization?

mpg replied:
Ok, so my memory wasn't that bad, I was just looking for the wrong thing: it's called `psa_crypto_init()` and I added it in the body of both test functions. Of course I agree it should go in the main test function, so this is a reminder for me to remove the existing occurrences when doing that.

mpg commented on include/mbedtls/pk.h:
```diff
@@ -232,6 +242,36 @@ void mbedtls_pk_restart_free( mbedtls_pk_restart_ctx *ctx );
+ * \return          #MBEDTLS_ERR_PK_BAD_INPUT_DATA on invalid input
+ *                  (context already used, invalid key slot)
+ * \return          #MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE if the key is not an
+ *                  ECC keypair,
```
I picked "keypair" that's the way it's currently called in `psa/crypto.h` - well actually while checking I just notice it's "key pair" with a space...

mpg replied:
Changing to "key pair" plus expanding on it in a parenthesis in the note above.

mpg commented:
@AndrzejKurek I've adjusted the names to use just "opaque" everywhere (rather than "opaque psa") for the reason described in the last commit, and for homogeneity with the name of the new API function in Hanno's #2156. I'm sorry this change comes after you've started working on using that for CSR creation, as that means you'll need to adjust you code.

For reference, here is the commands I used to make those changes, which you can probably use for your code too - the first 2 should be enough.
```
sed -i s/mbedtls_pk_setup_psa/mbedtls_pk_setup_opaque/g include/mbedtls/*.h library/*.c tests/suites/test_suite_*.function
sed -i s/MBEDTLS_PK_OPAQUE_PSA/MBEDTLS_PK_OPAQUE/g include/mbedtls/*.h library/*.c tests/suites/test_suite_*.function
sed -i s/pk_psa_/pk_opaque_/g library/pk_wrap.c
sed -i s/mbedtls_pk_opaque_psa/mbedtls_pk_opaque/g include/mbedtls/*.h library/*.c tests/suites/test_suite_*.function
```

@hanno-arm Thanks for your review, I'll address your feedback later this week.

mpg commented on include/mbedtls/pk.h:
```diff
@@ -232,6 +242,36 @@ void mbedtls_pk_restart_free( mbedtls_pk_restart_ctx *ctx );
+/**
+ * \brief           Initialize a PK context to wrap a PSA key slot.
+ *
+ * \param ctx       Context to initialize. Must be empty (type NONE).
```
Fixing.

mpg commented on include/mbedtls/pk.h:
```diff
@@ -232,6 +242,36 @@ void mbedtls_pk_restart_free( mbedtls_pk_restart_ctx *ctx );
+ *                  key slot might then be independently used or destroyed.
+ *
+ * \return          \c 0 on success,
+ * \return          #MBEDTLS_ERR_PK_BAD_INPUT_DATA on invalid input
```
Moving to full stops everywhere, as it's easier to manage.

mpg commented on include/mbedtls/pk.h:
```diff
@@ -232,6 +242,36 @@ void mbedtls_pk_restart_free( mbedtls_pk_restart_ctx *ctx );
+ *                  ECC keypair,
+ * \return          #MBEDTLS_ERR_PK_ALLOC_FAILED on allocation failure.
+ *
+ * \note            This function replaces mbedtls_pk_setup() for contexts
```
Moving right after the `\brief` section.

mpg commented on include/mbedtls/pk.h:
```diff
@@ -232,6 +242,36 @@ void mbedtls_pk_restart_free( mbedtls_pk_restart_ctx *ctx );
+ *                  that wrap a (possibly opaque) PSA key slot instead of
+ *                  storing and manipulating the key material directly.
+ *
+ * \note            This function is currently only available for ECC keypair.
```
Changing to "key pairs" while at it.

mpg commented:
@hanno-arm Thanks for your review. I believe I've addressed the pars of your feedback that didn't require rebasing over #2184. You may either review those changes now or wait until I rebased - I'll keep the pre-rebase version around anyway.

mpg commented:
@hanno-arm @AndrzejKurek I've addressed Hanno's feedback and rebased on #2184. Please review again. For reference, the branch right before rebasing is still available here: https://github.com/mpg/mbedtls/tree/iotssl-2580-pk-opaque-psa-0 Also, I've sligthly simplified the history after rebasing.

hanno-arm commented on include/mbedtls/ecdsa.h:
```diff
@@ -35,25 +35,30 @@
-#if MBEDTLS_ECP_MAX_BYTES > 124
-#error "MBEDTLS_ECP_MAX_BYTES bigger than expected, please fix MBEDTLS_ECDSA_MAX_LEN"
-#endif
+#define MBEDTLS_ECDSA_MAX_SIG_LEN( bits )                               \
```
I vaguely remember having seen this macro somewhere before - do we have other PRs introducing it, too, or did you write it from scratch, @mpg?

mpg replied:
The link's in the commit message: https://github.com/ARMmbed/mbedtls/pull/1293/files#diff-568ef321d275f2035b8b26a70ee9af0bR71 That's an old PR from a previous approach to opaque keys. Apparently it ran into a similar problem as this one with our current API not knowing about the buffer size. (The other PR that introduced it is dead, so introducing it here will not cause conflicts.)

hanno-arm replied:
I'm sorry, I missed the commit message.

mpg commented on include/mbedtls/ecdsa.h:
```diff
@@ -35,25 +35,30 @@
- *    twice that + 1 (tag) + 2 (len) for the sequence
- * (assuming ECP_MAX_BYTES is less than 126 for r and s,
- * and less than 124 (total len <= 255) for the sequence)
+ * For each of r and s, the value (V) may include an extra initial "0" bit.
```
Because ASN.1 says numbers that start with the msb set are negative, so when encoding positive numbers we should make sure they start with a `0` bit. See for example [here](https://docs.microsoft.com/en-us/windows/desktop/seccertenroll/about-integer) (I don't think there are public versions of the actual spec, but I think the MS docs should be quite reliable on this, and they coincide with everything's I've read so far).

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -761,20 +761,52 @@ static int pk_opaque_can_do( mbedtls_pk_type_t type )
 
-/* Like mbedtls_asn1_write_mpi, but from a buffer */
+/* translate PSA errors to best PK approximation */
+static int pk_err_from_psa( psa_status_t status )
```
Shall I move that to the utility PR?

mpg replied:
I don't know. It's specific to PK (as PSA errors are translated to the closest PK error), but it could still need to be shared between multiple files in the PK module. @AndrzejKurek what do you think? Would this function be useful in one of your PRs or tasks?

mpg replied:
@hanno-arm Actually I changed my mind while [reviewing another PR](https://github.com/ARMmbed/mbedtls/pull/2159#discussion_r233758471) and now think this should be moved to #2184 :)

hanno-arm replied:
@mpg Ok, I will do that and ping you once it's done.

hanno-arm replied:
@mpg Ping, I updated #2184 now.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -761,20 +761,52 @@ static int pk_opaque_can_do( mbedtls_pk_type_t type )
+ *
+ * p: pointer to the end of the output buffer
+ * start: start of the output buffer, and also of the mpi to write at the end
+ * n_len: length ot the mpi to read from start
```
Typo: `ot` -> `of`

Minor: The current documentation feels a bit too short to get what's happening. Suggestion:
```
/*
 * Simultaneously convert and move raw MPI from the beginning of a buffer
 * to an ASN.1 MPI at the end of the buffer.
 *
 * PARAM DESCRIPTION
 */
```

mpg replied:
Thanks for the suggestion! Using your text.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -716,4 +724,177 @@ const mbedtls_pk_info_t mbedtls_rsa_alt_info = {
+    *p -= len;
+    memmove( *p, start, len );
+
+    /* ASN.1 DER encoding requires minimal length, so skip leading 0s.
```
I don't feel comfortable with this precondition - in the end, PSA is an external module, and I think we should consider adding some sanity checks for what it returns to us. In this particular case, I think we should fail cleanly when given a buffer full of `0`s (maybe some PSA implementation at some point will erroneously return success instead of failure after zeroing the target buffer but not actually performing the operation? Or we miss do catch the error) by never allowing `len` to underflow.

mpg commented:
@hanno-arm Aw, excellent point, sorry for that. Hopefully fixed and ready for you to review now.

AndrzejKurek commented on tests/suites/test_suite_pk.function:
```diff
@@ -90,28 +90,64 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
+
+#include "mbedtls/psa_util.h"
+
+#define PK_PSA_INVALID_SLOT 0 /* guaranteed invalid */
```
How about adding such define in the PSA utils? Just a thought, not intended for this PR. It's used in several places and different PR's though.

hanno-arm commented on tests/suites/test_suite_pk.function:
```diff
@@ -90,13 +90,122 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
+ * or PK_PSA_INVALID_SLOT if no slot was available.
+ * The key uses NIST P-256 and is usable for signing with SHA-256.
+ */
+psa_key_slot_t pk_psa_genkey( void )
```
Please change the name of the function to indicate that it generates an EC key.

mpg replied:
In the future that function might be expanded to generate other types of keys, so for now I think this name's OK.

hanno-arm replied:
Yes

hanno-arm commented on tests/suites/test_suite_pk.data:
```diff
@@ -23,6 +23,9 @@ PK utils: ECDSA
 depends_on:MBEDTLS_ECDSA_C:MBEDTLS_ECP_DP_SECP192R1_ENABLED
 pk_utils:MBEDTLS_PK_ECDSA:192:24:"ECDSA"
 
+PK PSA utils
```
Please use a descriptive title for this test.

mpg commented:
@hanno-arm @AndrzejKurek I've rebased on recent #2206, make a small improvement suggested by Hanno, and extracted a utility function from #2183 Please review again.

mpg commented on tests/suites/test_suite_pk.data:
```diff
@@ -14,7 +14,7 @@ PK utils: ECDSA
 pk_utils:MBEDTLS_PK_ECDSA:192:24:"ECDSA"
 
-PK PSA utils
+PK PSA utililities: setup/free, info functions, unsupported operations
```
Fixed locally. I'm not pushing just for that, trying to preserve CI resources. Hopefully tomorrow I'll push a version rebased on the submodule anyway :)

mpg commented:
@hanno-arm @AndrzejKurek I've rebased on the rebased utilities branch. Please review again.

mpg commented:
Note: CI failures only due to known issues (GnuTLS fragmentation bug, timing test).

mpg commented:
Re-applying "DO NOT MERGE" as it needs an updated submodule before merging.
